README.md

```diff
@@ -64,6 +64,23 @@ At boot time, you might need `adb shell setenforce 0` for B2G to boot (flash_sar
   
 If need to output a zip ROM file, you can use `./build.sh dist DIST_DIR=dist_output` instead of `./build.sh` in step 5.
 
+## B2G Generic System Images (B2G-GSI) 
+More [detail](https://github.com/phhusson/treble_experimentations/wiki) of which device is supported and which type for your device.  
+1. Fetch the code: `REPO_INIT_FLAGS="--depth=1" ./config.sh b2g_gsi`
+2. Setup your environment to fetch the custom NDK: `export LOCAL_NDK_BASE_URL='ftp://ftp.kaiostech.com/ndk/android-ndk'`
+3. Install Gecko dependencies: `cd gecko && ./mach bootstrap`, choose option 1 (Boot2Gecko).
+4. Apply patch: `./patcher/patcher.sh`
+   - NOTE：Each patch is essence. If find one failed, do `repo sync` rollback HEAD. 
+5. Build: `./build-gsi.sh [your_device_type_name] systemimage`
```

Comment:
```
$ ./build-gsi.sh gsi_arm64_ab systemimage
[...]
FAILED: out/soong/.intermediates/system/libhidl/libhidlbase/android_arm64_armv8-a_vendor_shared/libhidlbase.so.abidiff
(prebuilts/clang-tools/linux-x86/bin/header-abi-diff -allow-unreferenced-changes -allow-unreferenced-elf-symbol-changes -lib libhidlbase -arch arm64 -o out/soong/.intermediates/system/libhidl/libhidlbase/android_arm64_armv8-a_vendor_shared/libhidlbase.so.abidiff -new out/soong/.intermediates/system/libhidl/libhidlbase/android_arm64_armv8-a_vendor_shared/libhidlbase.so.lsdump -old prebuilts/abi-dumps/vndk/29/64/arm64_armv8-a/source-based/libhidlbase.so.lsdump)|| (echo 'error: Please update ABI references with: $ANDROID_BUILD_TOP/development/vndk/tools/header-checker/utils/create_reference_dumps.py  -l libhidlbase' && (mkdir -p $DIST_DIR/abidiffs && cp out/soong/.intermediates/system/libhidl/libhidlbase/android_arm64_armv8-a_vendor_shared/libhidlbase.so.abidiff $DIST_DIR/abidiffs/) && exit 1)
******************************************************
error: VNDK library: libhidlbase's ABI has INCOMPATIBLE CHANGES Please check compatibility report at: out/soong/.intermediates/system/libhidl/libhidlbase/android_arm64_armv8-a_vendor_shared/libhidlbase.so.abidiff
******************************************************
error: Please update ABI references with: $ANDROID_BUILD_TOP/development/vndk/tools/header-checker/utils/create_reference_dumps.py  -l libhidlbase
[ 89% 68800/76929] //system/libhidl:libhidlbase header-abi-diff libhidlbase.so.abidiff [arm]
FAILED: out/soong/.intermediates/system/libhidl/libhidlbase/android_arm_armv8-a_vendor_shared/libhidlbase.so.abidiff
(prebuilts/clang-tools/linux-x86/bin/header-abi-diff -allow-unreferenced-changes -allow-unreferenced-elf-symbol-changes -lib libhidlbase -arch arm -o out/soong/.intermediates/system/libhidl/libhidlbase/android_arm_armv8-a_vendor_shared/libhidlbase.so.abidiff -new out/soong/.intermediates/system/libhidl/libhidlbase/android_arm_armv8-a_vendor_shared/libhidlbase.so.lsdump -old prebuilts/abi-dumps/vndk/29/64/arm_armv8-a/source-based/libhidlbase.so.lsdump)|| (echo 'error: Please update ABI references with: $ANDROID_BUILD_TOP/development/vndk/tools/header-checker/utils/create_reference_dumps.py  -l libhidlbase' && (mkdir -p $DIST_DIR/abidiffs && cp out/soong/.intermediates/system/libhidl/libhidlbase/android_arm_armv8-a_vendor_shared/libhidlbase.so.abidiff $DIST_DIR/abidiffs/) && exit 1)
******************************************************
error: VNDK library: libhidlbase's ABI has INCOMPATIBLE CHANGES Please check compatibility report at: out/soong/.intermediates/system/libhidl/libhidlbase/android_arm_armv8-a_vendor_shared/libhidlbase.so.abidiff
******************************************************
error: Please update ABI references with: $ANDROID_BUILD_TOP/development/vndk/tools/header-checker/utils/create_reference_dumps.py  -l libhidlbase
10:19:25 ninja failed with: exit status 1
make: *** [build/make/core/main.mk:21 : run_soong_ui] Erreur 1
```

Reply:
Fixed it - [Add SKIP_ABI_CHECKS to GSI](https://github.com/b2gOS/B2G/pull/15/commits/8fe49b50e4cf2c6d3de26267be5bea410cebb18c)

Reply:
Isnt it a bit dangerous? if there this kind of check I feel like we are going to have issues, nope?

Reply:
this issue is worth discussing.
I hope to support both A-only and A/B devices. However, due to historical reasons, type of A-only devices only exist on Android 8, which means that some compatibility patches need to be used, which may not pass the check.
In addition, if give up support for type A-only devices, this may improve.

Reply:
> this issue is worth discussing.
> I hope to support both A-only and A/B devices. However, due to historical reasons, type of A-only devices only exist on Android 8, which means that some compatibility patches need to be used, which may not pass the check.
> In addition, if give up support for type A-only devices, this may improve.

B2G only cares about Android 10+ :)

Reply:
haha. If the device supports GSI, but which Android version the device originally is, they can all run android 10 through GSI easily.

Reply:
Do you mean to give up A-only?
In this way, the significance of B2G-GSI is not too great :(. Many of the existing devices are lower than Android 10.

Reply:
I have to admit I have not looked into the details of A/B partitions, as well as of GSI, so I'm unable to understands the roots here.